opm/autodiff/ParallelRestrictedAdditiveSchwarz.hpp

```diff
@@ -0,0 +1,296 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef OPM_PARALLELRESTRICTEDADDITIVESCHWARZ_HEADER_INCLUDED
+#define OPM_PARALLELRESTRICTEDADDITIVESCHWARZ_HEADER_INCLUDED
+
+#include <dune/istl/preconditioner.hh>
+#include <dune/istl/paamg/smoother.hh>
+namespace Opm
+{
+template<class X, class Y, class C, class T>
+class ParallelRestrictedOverlappingSchwarz;
+template<class M, class X, class Y, class C>
+class ParallelOverlappingILU0;
+}
+namespace Dune
+{
+namespace Amg
+{
+template<class X, class Y, class C, class T>
+struct ConstructionTraits<Opm::ParallelRestrictedOverlappingSchwarz<X,Y,C,T> >
+{
+    typedef DefaultParallelConstructionArgs<T,C> Arguments;
+    typedef ConstructionTraits<T> SeqConstructionTraits;
+    static inline Opm::ParallelRestrictedOverlappingSchwarz<X,Y,C,T>* construct(Arguments& args)
+    {
+        return new Opm::ParallelRestrictedOverlappingSchwarz<X,Y,C,T>(*SeqConstructionTraits::construct(args),
+                                                                      args.getComm());
+    }
+
+    static inline void deconstruct(Opm::ParallelRestrictedOverlappingSchwarz<X,Y,C,T>* bp)
+    {
+        SeqConstructionTraits::deconstruct(static_cast<T*>(&bp->preconditioner));
+        delete bp;
+    }
+
+};
+template<class X, class Y, class C, class T>
+struct SmootherTraits<Opm::ParallelRestrictedOverlappingSchwarz<X,Y,C,T> >
+{
+    typedef DefaultSmootherArgs<typename T::matrix_type::field_type> Arguments;
+
+};
+
+template<class M, class X, class Y, class C>
+struct ConstructionTraits<Opm::ParallelOverlappingILU0<M,X,Y,C> >
+{
+    typedef Dune::SeqILU0<M,X,Y> T;
+    typedef DefaultParallelConstructionArgs<T,C> Arguments;
+    typedef ConstructionTraits<T> SeqConstructionTraits;
+    static inline Opm::ParallelOverlappingILU0<M,X,Y,C>* construct(Arguments& args)
+    {
+        return new Opm::ParallelOverlappingILU0<M,X,Y,C>(args.getMatrix(),
+                                                         args.getComm(),
+                                                         args.getArgs().relaxationFactor);
+    }
+
+    static inline void deconstruct(Opm::ParallelOverlappingILU0<M,X,Y,C>* bp)
+    {
+        delete bp;
+    }
+
+};/*
+    template<class X, class Y, class C>
+    struct SmootherTraits<Opm::ParallelOverlappingILU0<X,Y,C> >
+    {
+    typedef DefaultSmootherArgs<typename T::matrix_type::field_type> Arguments;
+
+    };*/
+
+}
+}
+
+namespace Opm{
+/**
+ * @brief Block parallel preconditioner.
+ *
+ * This is essentially a wrapper that takes a sequential
+ * preconditioner. In each step the sequential preconditioner
+ * is applied and then all owner data points are updated on
+ * all other processes.
+ */
+template<class X, class Y, class C, class T=Dune::Preconditioner<X,Y> >
```

Comment:
I always get confused by the nondescriptive template argument names. Would in not be better to use SequentialPreconditioner instead of T for example? (Assuming I understood that one...)